### PR TITLE
[Bifrost] ReplicatedLoglet read stream

### PR DESCRIPTION
**Pull request by AhmedSoliman**
---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/2028).
* #2039
* __->__ #2028

### CODE REVIEW

File: crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs [hunk 1]
```diff
@@ -0,0 +1,491 @@
...
+    let mut new_nodeset: Vec<_> = nodeset.iter().cloned().collect();
+    // Shuffle nodes
+    new_nodeset.shuffle(&mut rand::thread_rng());
+    // put my node at the end if it's there
+    if has_my_node_id {
+        new_nodeset.retain(|&x| x != my_node_id);
+        new_nodeset.push(my_node_id);
+    }
+
+    new_nodeset
```
Comment (tillrohrmann) [inline crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs:490]: Micro-optimization:

```
    let mut new_nodeset: Vec<_> = nodeset.iter().cloned().collect();
    // Shuffle nodes
    new_nodeset.shuffle(&mut rand::thread_rng());

    let has_my_node_idx = nodeset.iter().position(|&x| x == my_node_id);

    // put my node at the end if it's there
    if let Some(idx) = has_my_node_idx {
        new_nodeset.swap(idx, new_nodeset.len() - 1);
    }

    new_nodeset
```

could avoid some unnecessary moving of other elements.

**Review by tillrohrmann — APPROVED** [review]
Great work @AhmedSoliman 🚀  The changes look good to me. As far as I can tell, I think the logic is sound. I left a few minor comments. +1 for merging :-)

File: crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+            .bifrost
+            .replicated_loglet
+            .log_server_rpc_timeout;
+        // clamped to the channel capacity, ideally, the same value should be used to configure the
+        // channel size.
+        let readahead_max = 100.min(self.tx.max_capacity());
+        debug_assert!(readahead_max <= u16::MAX.into());
+        // This is automatically capped. This is the minimum number of slots that needs to be
+        // available in order to fetch a new batch.
+        let readahead_trigger_watermark = 50;
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs:134]: maybe even `readahead_trigger` is sufficient

File: crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs [hunk 3]
```diff
@@ -0,0 +1,491 @@
...
+            // - capacity is 100, watermark is 50; we reserve 100; but if watermark_max is 80, we
+            // request 80;
+            // - capacity is 5, watermark_min is 50; we wait until 50 is available.
+            let mut permits = self
+                .tx
+                .reserve_many(
+                    self.tx
+                        .capacity()
+                        .max(readahead_trigger_watermark)
+                        .min(readahead_max),
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/bifrost/src/providers/replicated_loglet/read_path/read_stream_task.rs:230]: Yes it's acceptable.

**Pull request merged**